# Snakefile

# --- Variables --- #
workdir: "/restricted/projectnb/cjdgenomics/CJD_GenomicsVariantCalling_20240430/"
names = ["BOSTON_CJD29_CJD29", "BOSTON_CJD40_CJD40"]  # Add your sample names here

# --- Pipeline Rules --- #
rule all:
    input:
        bam_processed = expand("results/{sample}/{sample}.aligned_fixmates_sorted.bai", sample = names),
        bam_index = expand("results/{sample}/{sample}.aligned_fixmates_sorted.bam", sample = names),
        dir = expand("results/{sample}",sample = names)
        # cannot use directory() in inputs

rule getreference:
    output:
        "reference/genome.fa",
        "reference/genome.fa.fai"
    shell:
        """
        wget -P reference/ https://dl.dnanex.us/F/D/734xvkZG02bbF5vF1bZ93G03ZK1g6x64ffVYkyBk/h38flat.fasta-index.tar.gz 
        tar -xvzf reference/h38flat.fasta-index.tar.gz
        """

rule create_directories:
    output:
        dir = directory(expand("results/{sample}",sample = names))
    shell:
        "mkdir -p {output.dir}"

rule cram_to_bam:
    input:
        cram="data/{sample}.aligned.cram",
        reference="reference/genome.fa"
    output:
        bam_processed="results/{sample}/{sample}.aligned_fixmates_sorted.bam",
        bam_index="results/{sample}/{sample}.aligned_fixmates_sorted.bai"
    params:
        memory="8G",
        threads=8
    shell:
        """
        samtools sort -n -@ {params.threads} -m {params.memory} -O BAM {input.cram} | \
        samtools fixmate -m -O BAM - - | \
        samtools sort -@ {params.threads} -m {params.memory} -O BAM -o {output.bam_processed} -
        samtools index {output.bam_processed}
        """
