# Snakefile

# --- Modules --- #
# module load htslib/1.19.1
# module load samtools/1.19.2

# --- Variables --- #
workdir: "/restricted/projectnb/cjdgenomics/CJD_GenomicsVariantCalling_20240430/"
samples = ["BOSTON_CJD29_CJD29","BOSTON_CJD40_CJD40"]  # Add your sample names here

# names = expand({name},name = samples)

# inputDataPath = path("/restricted/projectnb/cjdgenomics/CJD_GenomicsVariantCalling_20240430/data")
# resultPath = path("/restricted/projectnb/cjdgenomics/CJD_GenomicsVariantCalling_20240430/results")
# referencePath = path("/restricted/projectnb/cjdgenomics/CJD_GenomicsVariantCalling_20240430/reference")
# resultQualityPath = path("/restricted/projectnb/cjdgenomics/CJD_GenomicsVariantCalling_20240430/results/QC")


rule all:
    input:
        bam_processed = [f"results/{sample}/{sample}.aligned_fixmates_sorted.bam" for sample in samples],
        bam_index = [f"results/{sample}/{sample}.aligned_fixmates_sorted.bai" for sample in samples]

# --- Download Rules --- #
## getreference: Downloads the reference used in performing this pipeline for the variables. 
rule getreference:
    output:
        "reference/genome.fa",
        "reference/genome.fa.fai"
    shell:
        """
        # Using Wget to download the reference genome used in the current run of the data
        wget -P reference/ https://dl.dnanex.us/F/D/734xvkZG02bbF5vF1bZ93G03ZK1g6x64ffVYkyBk/h38flat.fasta-index.tar.gz 
        tar -xvzf reference/h38flat.fasta-index.tar.gz
        # samtools faidx genome.fa genome.fa.fai
        """

# --- Pipeline Rules --- #
## create_directories: Creates the directory for the results of various samples. 
rule create_directories:
    output:
        directory(f"results/{sample}" for sample in samples)
    shell:
        """
        mkdir -p {output}
        """

## cram_to_bam: Performs conversion to bam,fixingmates,sorting and creating bam index 
rule cram_to_bam:
    input:
        cram = [f"data/{sample}.aligned.cram" for sample in samples],
        reference = "reference/genome.fa"
    output:
        bam_processed = [f"results/{sample}/{sample}.aligned_fixmates_sorted.bam" for sample in samples],
        bam_index = [f"results/{sample}/{sample}.aligned_fixmates_sorted.bai" for sample in samples]
    params:
        memory = "8G",
        threads = "8",
        filetype = "BAM",
        
    shell:
        """
        samtools sort -n -@ {params.threads} -m {params.memory} -O {params.filetype} {input.cram} | \
        samtools fixmate -m -O {params.filetype} - - | \
        samtools sort -@ {params.threads} -m {params.memory} -O {params.filetype} -o {output.bam_processed}

        samtools index {output.bam_processed}
        """





# rule GATK_preprocessing:
#     input:
#     output:
#     shell:

# --- Notes --- #
#1) Using f"" for snakemake varaibles enclose them in double {{}}. 
#2) When using shell commands, do not use f"".

# Target Rule. Remember according to Snakemake syntax it should always be at the first so that when combining several workflows together it can be done properly. 