# Snakefile

# --- Modules --- #
# module load htslib/1.19.1
# module load samtools/1.19.2

# --- Variables --- #
samples = ["BOSTON_CJD29_CJD29","BOSTON_CJD40_CJD40"]  # Add your sample names here
#When scaling up use the samples filename
# with open(sample_names_file) as f:
#     samples = [line.strip() for line in f]

inputDataPath = "/restricted/projectnb/cjdgenomics/CJD_GenomicsVariantCalling_20240430/data"
resultPath = "/restricted/projectnb/cjdgenomics/CJD_GenomicsVariantCalling_20240430/results"
referencePath = "/restricted/projectnb/cjdgenomics/CJD_GenomicsVariantCalling_20240430/reference"
resultQualityPath = "/restricted/projectnb/cjdgenomics/CJD_GenomicsVariantCalling_20240430/results/QC"

rule all:
    input:
        expand(f"{resultPath}/{sample}/{sample}.aligned_fixmates_sorted.bam", sample=samples, resultPath = resultPath)



# --- Download Rules --- #
## getreference: Downloads the reference used in performing this pipeline for the variables. 
rule getreference:
    output:
        "genome.fa",
        "genome.fa.fai"
    shell:
        f"""
        # Using Wget to download the reference genome used in the current run of the data
        wget -P {referencePath} https://dl.dnanex.us/F/D/734xvkZG02bbF5vF1bZ93G03ZK1g6x64ffVYkyBk/h38flat.fasta-index.tar.gz 
        tar -xvzf {referencePath}/h38flat.fasta-index.tar.gz
        # samtools faidx genome.fa genome.fa.fai
        """

# --- Pipeline Rules --- #
## create_directories: Creates the directory for the results of various samples. 
rule create_directories:
    output:
        directory("{resultPath}/{sample}")
    shell:
        """
        mkdir -p {output}
        """

## cram_to_bam: Performs conversion to bam,fixingmates,sorting and creating bam index 
rule cram_to_bam:
    input:
        cram = f"{inputDataPath}/{sample}.aligned.cram",
        reference = f"{referencePath}/genome.fa"
    output:
        bam_processed = f"{resultPath}/{sample}/{sample}.aligned_fixmates_sorted.bam"
        bam_index = f"{resultPath}/{sample}/{sample}.aligned_fixmates_sorted.bai"
    params:
        memory = "8G"
        threads = "8"
        filetype = "BAM"
    shell:
    """
    samtools sort -n -@ {params.threads} -m {params.memory} -O {params.filetype} {input.cram} | \
    samtools fixmate -m -O {params.filetype} - - | \
    samtools sort -@ {params.threads} -m {params.memory} -O {params.filetype} -o {output.bam_processed}

    samtools index {output.bam_processed}
    """

# rule GATK_preprocessing:
#     input:
#     output:
#     shell:

# --- Notes --- #
#1) Using f"" for snakemake varaibles enclose them in double {{}}. 
#2) When using shell commands, do not use f"".

# Target Rule. Remember according to Snakemake syntax it should always be at the first so that when combining several workflows together it can be done properly. 